### PR TITLE
feat(v2): parallel batch processing

### PR DESCRIPTION
**Pull request by jeromevdl**
**Issue #, if available:** #1540

## Description of changes:

batch module can now process items in parallel rather than in sequence. Introduction of a new method `processBatchInParallel` in `BatchMessageHandler`. Works with SQS, Kinesis, DDB Streams but not SQS FIFO (does not make sense).

**Checklist**

 

* [x] [Meet tenets criteria](https://docs.powertools.aws.dev/lambda-java/#tenets)
* [x] Update tests
* [x] Update docs
* [x] PR title follows [conventional commit semantics](https://www.conventionalcommits.org/en/v1.0.0/)

## Breaking change checklist

**RFC issue #**:

* [ ] Migration process documented
* [ ] Implement warnings (if it can live side by side)

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Review by scottgerring — COMMENTED** [review]
Nice! Some feedback inline

File: docs/utilities/batch.md [hunk 1]
```diff
@@ -475,6 +466,46 @@ see the details for [SQS](https://docs.aws.amazon.com/lambda/latest/dg/with-sqs.
         }
     ```
 
+## Parallel processing
+You can choose to process batch items in parallel using the `BatchMessageHandler#processBatchInParallel()` 
+instead of `BatchMessageHandler#processBatch()`. Partial batch failure works the same way but items are processed
+in parallel rather than sequentially. 
+
+This feature is available for SQS, Kinesis and DynamoDB Streams but cannot be 
+used with SQS FIFO. In that case, items will be processed sequentially, even with the `processBatchInParallel` method.
```
[the remark this answers is not in the document]
Comment (jeromevdl) [reply docs/utilities/batch.md:475]: There is a log warn... 

Not sure about the behaviour we should have here.

File: docs/utilities/batch.md [hunk 2]
```diff
@@ -475,6 +466,46 @@ see the details for [SQS](https://docs.aws.amazon.com/lambda/latest/dg/with-sqs.
...
+
+!!! warning
+    Note that parallel processing is not always better than sequential processing, 
+    and you should benchmark your code to determine the best approach for your use case. 
+
+!!! info
+    To get more threads available (more vCPUs), you need to increase the amount of memory allocated to your Lambda function.
+    While the exact vCPU allocation isn't published, from observing common patterns customers see an allocation of one vCPU per 1024 MB of memory.
```
[the remark this answers is not in the document]
Comment (jeromevdl) [reply docs/utilities/batch.md:483]: if you don't have multiple vCPUs (threads), it won't be parallel. For example with 512 Mb of RAM, I think you get only one... To get 2, you need ~832Mb.
I can simplify the wording...

Comment (jeromevdl) [inline docs/utilities/batch.md:483]:
```suggestion
```

Comment (jeromevdl) [reply docs/utilities/batch.md:483]: OK with that @scottgerring ?

File: powertools-batch/src/main/java/software/amazon/lambda/powertools/batch/handler/DynamoDbBatchMessageHandler.java [hunk 3]
```diff
@@ -46,35 +49,60 @@ public DynamoDbBatchMessageHandler(Consumer<DynamodbEvent.DynamodbStreamRecord>
 
     @Override
     public StreamsEventResponse processBatch(DynamodbEvent event, Context context) {
-        List<StreamsEventResponse.BatchItemFailure> batchFailures = new ArrayList<>();
+        StreamsEventResponse response = StreamsEventResponse.builder().withBatchItemFailures(new ArrayList<>()).build();
 
-        for (DynamodbEvent.DynamodbStreamRecord record : event.getRecords()) {
-            try {
+        for (DynamodbEvent.DynamodbStreamRecord streamRecord : event.getRecords()) {
```
[the remark this answers is not in the document]
Comment (jeromevdl) [reply powertools-batch/src/main/java/software/amazon/lambda/powertools/batch/handler/DynamoDbBatchMessageHandler.java:54]: you're right, it's because I started with SQS which required to be written differently and then used the same logic...

File: powertools-batch/src/main/java/software/amazon/lambda/powertools/batch/handler/DynamoDbBatchMessageHandler.java [hunk 4]
```diff
@@ -46,35 +49,60 @@ public DynamoDbBatchMessageHandler(Consumer<DynamodbEvent.DynamodbStreamRecord>
...
+
+
+    @Override
+    public StreamsEventResponse processBatchInParallel(DynamodbEvent event, Context context) {
+        MultiThreadMDC multiThreadMDC = new MultiThreadMDC();
+
+        List<StreamsEventResponse.BatchItemFailure> batchItemFailures = event.getRecords()
+                .parallelStream() // Parallel processing
+                .map(eventRecord -> {
+                    multiThreadMDC.copyMDCToThread(Thread.currentThread().getName());
```
[the remark this answers is not in the document]
Comment (jeromevdl) [reply powertools-batch/src/main/java/software/amazon/lambda/powertools/batch/handler/DynamoDbBatchMessageHandler.java:69]: What is using ThreadLocal in tracing ?

Serialization yes, with this ObjectMapper that we need to think about seriously. Tracing, I'm not sure...

File: powertools-batch/src/main/java/software/amazon/lambda/powertools/batch/handler/SqsBatchMessageHandler.java [hunk 5]
```diff
@@ -61,57 +66,27 @@ public SQSBatchResponse processBatch(SQSEvent event, Context context) {
         // If we are working on a FIFO queue, when any message fails we should stop processing and return the
         // rest of the batch as failed too. We use this variable to track when that has happened.
         // https://docs.aws.amazon.com/lambda/latest/dg/with-sqs.html#services-sqs-batchfailurereporting
-        boolean failWholeBatch = false;
+        AtomicBoolean failWholeBatch = new AtomicBoolean(false);
```
[the remark this answers is not in the document]
Comment (jeromevdl) [reply powertools-batch/src/main/java/software/amazon/lambda/powertools/batch/handler/SqsBatchMessageHandler.java:69]: see below

File: powertools-batch/src/main/java/software/amazon/lambda/powertools/batch/handler/SqsBatchMessageHandler.java [hunk 6]
```diff
@@ -61,57 +66,27 @@ public SQSBatchResponse processBatch(SQSEvent event, Context context) {
...
 
         int messageCursor = 0;
-        for (; messageCursor < event.getRecords().size() && !failWholeBatch; messageCursor++) {
+        for (; messageCursor < event.getRecords().size() && !failWholeBatch.get(); messageCursor++) {
```
[the remark this answers is not in the document]
Comment (jeromevdl) [reply powertools-batch/src/main/java/software/amazon/lambda/powertools/batch/handler/SqsBatchMessageHandler.java:72]: it is because of this: 

`processBatchItem(message, context).ifPresent(batchItemFailure -> {`

I could not keep the simple boolean...

[the remark this answers is not in the document]
Comment (jeromevdl) [reply powertools-batch/src/main/java/software/amazon/lambda/powertools/batch/handler/SqsBatchMessageHandler.java:72]: (needs the final for the stream)

File: powertools-batch/src/main/java/software/amazon/lambda/powertools/batch/handler/SqsBatchMessageHandler.java [hunk 7]
```diff
@@ -121,4 +96,60 @@ public SQSBatchResponse processBatch(SQSEvent event, Context context) {
         }
         return response;
     }
+
+    @Override
+    public SQSBatchResponse processBatchInParallel(SQSEvent event, Context context) {
```
[the remark this answers is not in the document]
Comment (jeromevdl) [reply powertools-batch/src/main/java/software/amazon/lambda/powertools/batch/handler/SqsBatchMessageHandler.java:101]: not sure, let's ask on slack

File: powertools-batch/src/main/java/software/amazon/lambda/powertools/batch/internal/MultiThreadMDC.java [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+import java.util.Map;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.slf4j.MDC;
+
+/**
+ * MDC (SLF4J) is not passed to other threads (ThreadLocal).
+ * This class permits to manually copy the MDC to a given thread.
+ */
+public class MultiThreadMDC {
```
[the remark this answers is not in the document]
Comment (jeromevdl) [reply powertools-batch/src/main/java/software/amazon/lambda/powertools/batch/internal/MultiThreadMDC.java:28]: I see your point but I don't want to make batch depends on logging...

File: powertools-batch/src/main/java/software/amazon/lambda/powertools/batch/handler/DynamoDbBatchMessageHandler.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (scottgerring) [reply powertools-batch/src/main/java/software/amazon/lambda/powertools/batch/handler/DynamoDbBatchMessageHandler.java:69]: Tracing will use it for span propagation I reckon

File: docs/utilities/batch.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (scottgerring) [reply docs/utilities/batch.md:475]: We agreed to throw here on the team chat now i reckon?

File: docs/utilities/batch.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (scottgerring) [reply docs/utilities/batch.md:483]: I don't think that's true. If you have a single vCPU and you block on IO the system scheduler will sub another process in to do work. Multithreading gives a benefit even in single core systems.

File: powertools-batch/src/main/java/software/amazon/lambda/powertools/batch/internal/MultiThreadMDC.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (scottgerring) [reply powertools-batch/src/main/java/software/amazon/lambda/powertools/batch/internal/MultiThreadMDC.java:28]: I see your point

File: docs/utilities/batch.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (scottgerring) [reply docs/utilities/batch.md:483]: You're right. the issue is that the `ForkJoinPool` defaults to [number of available processors](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ForkJoinPool.html) threads. A thread will block on IO, so if we have 1 vCPU, we're going to run one thread for the pool, it'll block, and the function will idle until the IO comes back.

I see two options:

* `java.util.concurrent.ForkJoinPool.common.parallelism` document turning this up in the case you know you are IO-bound. If you're not doing CPU dominant stuff this'll _just work_
* Switch the batch interface up to use `CompletableFuture`s. This is a bigger piece of work, and I think would only work if the code the user wrote chained future's together, rather than doing blocking IO. So maybe not much help?

I feel like the former is probably better. Simpler and less burden on dev. Thoughts @jeromevdl ? There's probably some options with reactive streams or so too, but I have no experience there and it feels like it's getting a bit wild.

File: powertools-batch/src/main/java/software/amazon/lambda/powertools/batch/handler/DynamoDbBatchMessageHandler.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (scottgerring) [reply powertools-batch/src/main/java/software/amazon/lambda/powertools/batch/handler/DynamoDbBatchMessageHandler.java:68]: X-Ray documents how we can explicitly [pass segment context between threads](https://docs.aws.amazon.com/xray/latest/devguide/xray-sdk-java.html#xray-sdk-java-multithreading). The trick'll be that we need to do this without binding the two libraries together.

File: powertools-batch/src/main/java/software/amazon/lambda/powertools/batch/handler/SqsBatchMessageHandler.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (scottgerring) [reply powertools-batch/src/main/java/software/amazon/lambda/powertools/batch/handler/SqsBatchMessageHandler.java]: There's also [AtomicBoolean](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/atomic/AtomicBoolean.html) - maybe a good one?

It turns out the `atomic` package is lockfree anyway so I think my concern is a bit unwarranted. I think the array form is a bit mysterious in comparison to using the package.

File: docs/utilities/batch.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (jeromevdl) [reply docs/utilities/batch.md:483]: This is what I've read at some point, could not remember where...

Also this: https://www.baeldung.com/java-when-to-use-parallel-stream#2-common-thread-pool
> It’s important to remember that this is a global setting and that it will affect all parallel streams and any other fork-join tasks that use the common pool. We strongly suggest that this parameter is not modified unless we have a very good reason for doing so.

We can probably speak about this setting, but also reinforce the message about testing because it can happen weird things: parallel processing (much) slower than sequential.

I don't think we should complexify the code, and devX with futures...

File: powertools-batch/src/main/java/software/amazon/lambda/powertools/batch/handler/SqsBatchMessageHandler.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (jeromevdl) [reply powertools-batch/src/main/java/software/amazon/lambda/powertools/batch/handler/SqsBatchMessageHandler.java]: > There's also [AtomicBoolean](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/atomic/AtomicBoolean.html) - maybe a good one?

Are you kidding? This is what I use...

File: powertools-batch/src/main/java/software/amazon/lambda/powertools/batch/handler/DynamoDbBatchMessageHandler.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (jeromevdl) [reply powertools-batch/src/main/java/software/amazon/lambda/powertools/batch/handler/DynamoDbBatchMessageHandler.java:68]: Good catch... but we probably don't want to have the X-Ray dependency (or even the PT tracing one) in batch.
Then how can we do ? Some dirty reflection:

```java
    Class<?> awsXRayClass =
                Class.forName("com.amazonaws.xray.AWSXRay", true, this.getClass().getClassLoader());
    Method getGlobalRecorder = awsXRayClass.getMethod("getGlobalRecorder");
    Object awsXRayRecorder = getGlobalRecorder.invoke(null);
    Method getTraceEntity = awsXRayRecorder.getClass().getMethod("getTraceEntity");
    Object segment = getTraceEntity.invoke(awsXRayRecorder);
    ...
```

File: docs/utilities/batch.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (scottgerring) [reply docs/utilities/batch.md:483]: Alternatively - how about on the builder we provide a mechanism to override the thread pool? That way if users want they can provide a customized version, and they can separate it from the common pool. This is more explicit and less "global magic" than the `-D` param. We have a nice place for it in the builder interfaces. Thoughts?

> I don't think we should complexify the code, and devX with futures...
Agreed. And if the futures do blocking stuff it doesn't improve the situation anyway.

File: powertools-batch/src/main/java/software/amazon/lambda/powertools/batch/handler/DynamoDbBatchMessageHandler.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (jeromevdl) [reply powertools-batch/src/main/java/software/amazon/lambda/powertools/batch/handler/DynamoDbBatchMessageHandler.java:68]: There is no such service thing for XRay, and even if we created one, we still need the XRay classes/interfaces...

It makes me think of this issue #1505 that @mriccia created. Should we introduce XRay in our code, even if we don't know if people use tracing? `aws-java-sdk-xray-1.12.228.jar` is 611 kb...

[the remark this answers is not in the document]
Comment (jeromevdl) [reply powertools-batch/src/main/java/software/amazon/lambda/powertools/batch/handler/DynamoDbBatchMessageHandler.java:68]: Based in my tests, I don't see any problem with Tracing:

image

Works both with the `@Tracing` annotation on processMessage method and with the `TracingUtils.withSubsegment("processMessage", (subsegment -> {...}));`

Could also add annotations and metadata.

[the remark this answers is not in the document]
Comment (scottgerring) [reply powertools-batch/src/main/java/software/amazon/lambda/powertools/batch/handler/DynamoDbBatchMessageHandler.java:68]: >  Should we introduce XRay in our code, even if we don't know if people use tracing?

I think no, because we are actively decoupling tracing from x-ray at the moment (#1468, @jasoniharris and I are working on this now). If anything we'd couple around some lightweight interface in `powertools-tracing`. Maybe that's our soln? 

> Works both with the @Tracing annotation on processMessage method and with the 

I'm not sure this shows it's working. Can you show me that the segments "outside of" the batch handler - which are definitely running in another thread - are tied together with these?

**Comment by humanzz** [comment]
Hi folks,

I'm quite happy to see you're working on this. I know my team would love such ability.

I have not read the PR yet, but thought I'd leave a couple of comments to consider - if you haven't already.

In some of our Lambda functions, where we use powertools (more specifically logging, tracing), we do leverage concurrency to speedup our lambda requests. Depending on the use case, this concurrency is achieved using (1) AWS SDK async clients (which return Completable futures) (2) Other types of clients that might have their thread pools (3) Explicit executor services (and sometimes virtual threads from Java 21).

The core challenges we tend to see are in logging/tracing
1. In the beginning of handling a request, we like to add some properties to both powertools logging context and metrics context. This allows us to easily debug issues, and trace requests, via business-relevant identifiers
2. We see challenges with xray tracing e.g. either no trace emitted, or trace having wrong parent. All of this is likely due to xray library depending on thread locals to have this lineage tracking, which becomes very challenging with existing thread pools we don't control. Most of the time we end up having to write some custom pieces of code to try to pass relevant info across threads when possible
3. We also see challenges with logging context, again with thread pools we don't control

I kinda feel, this change, is likely the first time that powertools is handling concurrency. I wonder, if as part of that, or as follow ups, challenges with concurrency when it comes to tracing/logging can maybe be looked at, to see if there's any level of help powertools can provide.

**Comment by scottgerring** [comment]
> We see challenges with xray tracing e.g. either no trace emitted, or trace having wrong parent. All of this is likely due to xray library depending on thread locals to have this lineage tracking, which becomes very challenging with existing thread pools we don't control. Most of the time we end up having to write some custom pieces of code to try to pass relevant info across threads when possible

Hey @humanzz ! Sorry for the slow reply; it's been busy over here.
@jeromevdl and I were speaking about this this morning. The error mode you describe from your experience with pooling - missing traces and strange parents - is the sort of thing I am a keen to see we avoid [here](https://github.com/aws-powertools/powertools-lambda-java/pull/1620#discussion_r1565434974). 

I think the broader issue of "things that need to propagate information across thread boundaries don't work well" is something we may well be positioned to help with - certainly with logging and tracing at least. I have opened #1670 to track this.

**Comment by scottgerring** [comment]
@jeromevdl how about we merge this without the fix for x-ray and handle that on another PR?

Given the x-ray and log correlation is likely going to be a bigger thing, and we are talking about a SNAPSHOT here, I think the downside of merging this into v2 is low - some x-ray traces won't correlate well on a preview release - and it will let us deal with the bigger issue asynchronously rather than scope-creeping this PR.

**Comment by jeromevdl** [comment]
> @jeromevdl how about we merge this without the fix for x-ray and handle that on another PR?
> 
> Given the x-ray and log correlation is likely going to be a bigger thing, and we are talking about a SNAPSHOT here, I think the downside of merging this into v2 is low - some x-ray traces won't correlate well on a preview release - and it will let us deal with the bigger issue asynchronously rather than scope-creeping this PR.

I agree, let's review like it is today and create a new issue for the multithread support

**Comment by jeromevdl** [comment]
See #1671

**Comment by scottgerring** [comment]
@jeromevdl i'll review this tomorrow morning

File: docs/utilities/batch.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (scottgerring) [reply docs/utilities/batch.md:475]: @jeromevdl coming back to this one - I believe we discussed this in slack with the rest of the PT gang and agreed that throwing was a better devExp - should we change?

File: docs/utilities/batch.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (scottgerring) [reply docs/utilities/batch.md:483]: @jeromevdl coming back to this one - should we let the user override the thread pool and have some brief docs about this?

File: powertools-batch/src/main/java/software/amazon/lambda/powertools/batch/handler/BatchMessageHandler.java [hunk 9]
```diff
@@ -33,6 +33,21 @@ public interface BatchMessageHandler<E, R> {
...
+    R processBatch(E event, Context context);
 
+    /**
+     * Processes the given batch in parallel returning a partial batch
+     * response indicating the success and failure of individual
+     * messages within the batch. <br/>
+     * Note that parallel processing is not always better than sequential processing,
+     * and you should benchmark your code to determine the best approach for your use case. <br/>
+     * Also note that to get more threads available (more vCPUs),
+     * you need to increase the amount of memory allocated to your Lambda function. <br/>
```
Comment (scottgerring) [inline powertools-batch/src/main/java/software/amazon/lambda/powertools/batch/handler/BatchMessageHandler.java:45]: Or, override the thread pool on the builder , if we add this

File: powertools-batch/src/main/java/software/amazon/lambda/powertools/batch/handler/DynamoDbBatchMessageHandler.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (scottgerring) [reply powertools-batch/src/main/java/software/amazon/lambda/powertools/batch/handler/DynamoDbBatchMessageHandler.java:68]: To wrap this thread up, we have seen that we get spans disconnecting from each other when throwing work into the ThreadPool. We've raised #1671 to deal with this asynchronously

File: docs/utilities/batch.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (jeromevdl) [reply docs/utilities/batch.md]: I changed the code but not doc, it's done now

File: docs/utilities/batch.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (jeromevdl) [reply docs/utilities/batch.md]: In both case (`-D`) or manually in the code, it is not a best practice to do this.
See https://www.baeldung.com/java-when-to-use-parallel-stream#2-common-thread-pool & https://www.baeldung.com/java-when-to-use-parallel-stream#3-custom-thread-pool. 

Also read this: https://dzone.com/articles/be-aware-of-forkjoinpoolcommonpool. While it looks great to enable some kind of customization, it can really downgrade performance... We can maybe document the -D option so that people who know what they do use it, but we should not simplify it too much as we will receive comments that parallelism works slower than sequential...

File: powertools-batch/src/main/java/software/amazon/lambda/powertools/batch/handler/BatchMessageHandler.java (hunk 9 above)
Comment (jeromevdl) [reply powertools-batch/src/main/java/software/amazon/lambda/powertools/batch/handler/BatchMessageHandler.java:45]: I think it's a bad idea...

Comment (scottgerring) [reply powertools-batch/src/main/java/software/amazon/lambda/powertools/batch/handler/BatchMessageHandler.java:45]: @jeromevdl can you elaborate?

Comment (jeromevdl) [reply powertools-batch/src/main/java/software/amazon/lambda/powertools/batch/handler/BatchMessageHandler.java:45]: https://github.com/aws-powertools/powertools-lambda-java/pull/1620#discussion_r1654331875

File: docs/utilities/batch.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (scottgerring) [reply docs/utilities/batch.md]: To be clear, I don't think we should customize the common pool, i think we should allow the user to provide a custom pool if they so desire. This opens up the ability to benefit from this PR for IO-bound tasks. The dzone link summarises this:

> if you intend to use commonPool for blocking tasks, then you need to consider some consequences. If you have more than three available CPUs, then your commonPool is automatically sized to two threads and you can very easily block execution of any other part of your system that uses the commonPool at the same time by keeping the threads in a blocked state. **As a rule of thumb, we can create our own thread-pool for blocking tasks and keep the rest of the system separated and predictable.**

If we don't want to do this, we need to make it explicit that the user should not use this change for operations that perform blocking IO in the docs. 

My thinking is 1/ Lambdas often do downstream IO-bound work that can be effectively parallelized but users will not be able to take advantage of this without a custom thread pool (or the `-D` and the flow on affects of that) and 2/ it's a pretty inoffensive escape hatch?

[the remark this answers is not in the document]
Comment (jeromevdl) [reply docs/utilities/batch.md]: Based on our discussion, we won't provide the option for custom thread pool for now. This feature can thus not be used for IO-bound work. We'll document it.

_Motivations_: 
- if we provide such a feature, some people might use it for any kind of job and expect performance to be better, while in most case it would be worse. It might gives the impression Powertools has a problem and people will open issues for this.
- By adding this feature we include a public api that would be difficult to remove in cases of miss-use (breaking change). On the other side, we might add this feature later on if people are willing to use it for IO-bound stuff and are conscious of what they do.

File: docs/utilities/batch.md [hunk 10]
```diff
@@ -475,6 +466,49 @@ see the details for [SQS](https://docs.aws.amazon.com/lambda/latest/dg/with-sqs.
...
+
+!!! warning
+    Note that parallel processing is not always better than sequential processing, 
+    and you should benchmark your code to determine the best approach for your use case. 
+
+!!! info
+    To get more threads available (more vCPUs), you need to increase the amount of memory allocated to your Lambda function.
+    While it is technically possible to increase the number of threads using Java options or custom thread pools, it can 
+    decrease performance when miss-used (see [here](https://www.baeldung.com/java-when-to-use-parallel-stream#fork-join-framework)
+    and [here](https://dzone.com/articles/be-aware-of-forkjoinpoolcommonpool) for more information) and we don't recommend it.
```
Comment (jeromevdl) [inline docs/utilities/batch.md:485]: @scottgerring feel free to suggest something else as I'm not sure what else you wanted to say. I'm not sure if we want to speak about IO-bound tasks really.

File: docs/utilities/batch.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (scottgerring) [reply docs/utilities/batch.md]: Thanks for summarising @jeromevdl , and as always a good chat :)

File: docs/utilities/batch.md [hunk 11]
```diff
@@ -480,6 +480,9 @@ used with SQS FIFO. In that case, an `UnsupportedOperationException` is thrown.
 
 !!! info
     To get more threads available (more vCPUs), you need to increase the amount of memory allocated to your Lambda function.
+    While it is technically possible to increase the number of threads using Java options or custom thread pools, it can 
+    decrease performance when miss-used (see [here](https://www.baeldung.com/java-when-to-use-parallel-stream#fork-join-framework)
+    and [here](https://dzone.com/articles/be-aware-of-forkjoinpoolcommonpool) for more information) and we don't recommend it.
```
Comment (scottgerring) [inline docs/utilities/batch.md:485]: how about this?

 It is possible to increase the number of threads using Java options or custom thread pools. In most cases the defaults work well, and changing them is more likely to decrease performance - see [here](https://www.baeldung.com/java-when-to-use-parallel-stream#fork-join-framework) and [here](https://dzone.com/articles/be-aware-of-forkjoinpoolcommonpool). In situations where this may be useful - such as performing IO-bound work in parallel - make sure to measure before and after!

**Review by scottgerring — APPROVED** [review]
🤝
